### PR TITLE
[flytepropeller] Watch agent metadata service dynamically

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue
https://github.com/flyteorg/flyte/issues/3936

## Why are the changes needed?
Users don't need to configure `supportedTaskTypes` for agent service after this PR merged.
We will connect to the agent server (or agent pod) to retrieve agent metadata every 10 seconds.

The config map for `agent-service` will be like:
```yaml
  agent-service:
    supportedTaskTypes:
      # - chatgpt
      # - sensor  
      # - spark
    #   - default_task
    #   - custom_task
    #   - sensor
    #   - airflow
    # By default, all the request will be sent to the default agent.
    defaultAgent:
      endpoint: "localhost:8000"
      insecure: true
      timeouts:
        CreateTask: 100s
        GetTask: 100s
      defaultTimeout: 100s
```

## Architecture Diagram
image

For example, the lifecycle in the propeller with agent will be like below
 
1. call the `watchAgent` function in `webapi/agent/plugin.go`
2. send request to `localhost:8000` and `localhost:8001` 
3. update  `agentRegistry` // map[taskTypeName][taskTypeVersion] => Agent
4. update `agentService` *core.AgentService
 // Propeller will use it to route to the correct plugin. In our case, it’s agent-service.
5. wait for 10 seconds `p.cfg.PollInterval.Duration`, and go back to 1

```yaml
agent-service:
    # By default, all the request will be sent to the default agent.
    defaultAgent:
      endpoint: "localhost:8000"
      insecure: true
      timeouts:
        CreateTask: 100s
        GetTask: 100s
      defaultTimeout: 100s
    agents:
      custom_agent:
        endpoint: "localhost:8001"
        insecure: true
        timeouts:
          ExecuteTaskSync: 300s
          GetTask: 100s
        defaultTimeout: 300s
    agentForTaskTypes:
      # It will override the default agent for custom_task, which means propeller will send the request to this agent.
      - chatgpt: custom_agent
      - airflow: custom_agent
```

## What changes were proposed in this pull request?
TBD

## How was this patch tested?
single binary

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flyte/issues/3936

### CODE REVIEW

**Comment by Future-Outlier** [comment]
@honnix sorry for the late reply and thank you for your patient.
After this PR is merged, the watcher is done, and you can build the image to update your propeller.

Thank you so much!

**Comment by honnix** [comment]
Thank you for working on the watcher feature! Could you please write some description to help me understand what this achieves? Thank you.

**Comment by Future-Outlier** [comment]
> Thank you for working on the watcher feature! Could you please write some description to help me understand what this achieves? Thank you.

No problem, I will do it, thank you!

**Comment by Future-Outlier** [comment]
Update: Facing issue with race condition, will try to solve it then we can merge it!!

**Comment by Future-Outlier** [comment]
Hi @EngHabu,

The agent watcher can now dynamically update the supported task types for `agent-service`.

Kevin and I discovered that we need to perform two tasks concurrently:
1. Update the supported task types for `agent-service`.
2. Check if the task type from TaskMetadata is in the supported task types.

Therefore, we need to use `mutex lock` in our code to handle these operations.

Thank you for reading this and have a nice weekend!

**Comment by Future-Outlier** [comment]
Hi, @honnix, can you take a look at this PR? 

The implementation might change, but can you help review it for any issues?
Thank you!

**Comment by honnix** [comment]
@Future-Outlier I didn't look into details of the change, but overall it's a great feature to be able to refresh agent capability at runtime without restarting propeller.

**Comment by Future-Outlier** [comment]
Hello, @EngHabu.
Can you take a look? Kevin and I will fix it ASAP if any changes are needed. Thank you!

File: flytepropeller/pkg/controller/nodes/task/handler.go [hunk 1]
```diff
@@ -337,6 +344,11 @@ func (t Handler) ResolvePlugin(ctx context.Context, ttype string, executionConfi
 		logger.Debugf(ctx, "Plugin [%s] resolved for Handler type [%s]", p.GetID(), ttype)
 		return p, nil
 	}
+
+	if t.agentService.ContainTaskType(ttype) {
+		return t.agentService.CorePlugin, nil
+	}
+
```
Comment (Future-Outlier) [inline flytepropeller/pkg/controller/nodes/task/handler.go:351]: Here is how we route to `agent-service` core plugin.

File: flytepropeller/pkg/controller/nodes/task/handler.go [hunk 2]
```diff
@@ -200,6 +201,7 @@ type Handler struct {
 	pluginScope     promutils.Scope
 	eventConfig     *controllerConfig.EventConfig
 	clusterID       string
+	agentService    *pluginCore.AgentService
```
Comment (Future-Outlier) [inline flytepropeller/pkg/controller/nodes/task/handler.go:204]: Here is important, we add a new field in this struct.

**Comment by Future-Outlier** [comment]
Hi, @honnix 
Before we merge this PR, can you help test this one in your scenario?
I can help you build the image you need, here are the actionable steps.

1. checkout this branch
2. Use this Dockerfile to build it, replace the flytepropeller's deployment and restart the flytepropeller
https://github.com/flyteorg/flyte/blob/master/Dockerfile.flytepropeller

File: flyteplugins/go/tasks/plugins/webapi/agent/client.go [hunk 3]
```diff
@@ -97,17 +97,11 @@ func getFinalContext(ctx context.Context, operation string, agent *Deployment) (
 	return context.WithTimeout(ctx, timeout)
 }
 
-func updateAgentRegistry(ctx context.Context, cs *ClientSet) {
-	agentRegistry := make(Registry)
+func getUpdatedAgentRegistry(ctx context.Context, cs *ClientSet) Registry {
```
Comment (pingsutw) [inline flyteplugins/go/tasks/plugins/webapi/agent/client.go:100]: nit:
```suggestion
func getAgentRegistry(ctx context.Context, cs *ClientSet) Registry {
```

Comment (Future-Outlier) [reply flyteplugins/go/tasks/plugins/webapi/agent/client.go]: FIxed, thank you, and I also added tests

File: flyteplugins/go/tasks/pluginmachinery/core/plugin.go [hunk 4]
```diff
@@ -55,7 +58,27 @@ type Plugin interface {
...
 
-// Loads and validates a plugin.
+type AgentService struct {
+	mu                 sync.RWMutex
+	supportedTaskTypes []TaskType
+	CorePlugin         Plugin
+}
+
+// ContainTaskType check if agent supports this task type.
+func (p *AgentService) ContainTaskType(taskType TaskType) bool {
```
[the remark this answers is not in the document]
Comment (Future-Outlier) [reply flyteplugins/go/tasks/pluginmachinery/core/plugin.go:68]: No problem, will do it now

**Comment by honnix** [comment]
> Hi, @honnix Before we merge this PR, can you help test this one in your scenario? I can help you build the image you need, here are the actionable steps.
> 
> 1. checkout this branch
> 2. Use this Dockerfile to build a new image, replace the flytepropeller's deployment and restart the flytepropeller
>    [`master`/Dockerfile.flytepropeller](https://github.com/flyteorg/flyte/blob/master/Dockerfile.flytepropeller?rgh-link-date=2024-06-14T04%3A18%3A46Z)

Sorry for replying late. I was on vacation.

I unfortunately don't have capacity to take on this. I can ask around in the team.

**Comment by Future-Outlier** [comment]
> > Hi, @honnix Before we merge this PR, can you help test this one in your scenario? I can help you build the image you need, here are the actionable steps.
> > 
> > 1. checkout this branch
> > 2. Use this Dockerfile to build a new image, replace the flytepropeller's deployment and restart the flytepropeller
> >    [`master`/Dockerfile.flytepropeller](https://github.com/flyteorg/flyte/blob/master/Dockerfile.flytepropeller?rgh-link-date=2024-06-14T04%3A18%3A46Z)
> 
> Sorry for replying late. I was on vacation.
> 
> I unfortunately don't have capacity to take on this. I can ask around in the team.

No problem, I will be on vacation too!
Have a nice vacation!

**Review by pingsutw — APPROVED** [review]

**Pull request merged**